File: Python/92.Reverse Linked List II.py
#把链表中第n到第m个位置之间逆序
class ListNode(object):
     def __init__(self, x):
         self.val = x
         self.next = None

class Solution(object):
    def reverseBetween(self, head, m, n):
        nHead = ListNode(0)
        nHead.next = head
        i , p , q = 0 , nHead , head
        while i<m-1 and q!=None:
            i , p  ,q = i+1 , p.next , q.next
        mm , p = p , q
        q = p.next
        while i<n-1:
            if q!=None:
                t = q.next
            q.next = p
            i , p , q = i+1 , q , t
        nn = mm.next
        nn.next = q
        mm.next = p
        return nHead.next


l1 = ListNode(1)
l2 = ListNode(2)
l3 = ListNode(3)
l4 = ListNode(4)
l5 = ListNode(5)
l6 = ListNode(6)
l1.next = l2
l2.next = l3
l3.next = l4
l4.next = l5
l5.next = l6

s = Solution()
p = s.reverseBetween(l1,1,6)
while p!=None:
    print(p.val)
    p = p.next

File: C++/Linked List/92.Reverse Linked List II.cpp
/*
把链表中第m和第n个结点之间逆序。
*/

#include <iostream>
using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
	ListNode* reverseBetween(ListNode* head, int m, int n) {
        ListNode nHead(0);
        nHead.next = head;
        ListNode *preM = &nHead, *cur,*tmp;
        for (int i = 0; i < m-1; ++i)
        {
        	preM = preM->next;
        }
        cur = preM->next;
        for(int i=0; i < n-m; ++i)
        {
        	tmp = cur->next;
        	cur->next = tmp->next;
        	tmp->next = preM->next;
        	preM->next = tmp;
        }
        return nHead.next;
    }
};
int main()
{
	Solution s;
	ListNode head(1);
	ListNode t1(2);
	ListNode t2(3);
	ListNode t3(4);
	ListNode t4(5);
	ListNode t5(6);
	head.next = &t1;
	t1.next = &t2;
	t2.next = &t3;
	t3.next = &t4;
	t4.next = &t5;
 	ListNode* h = s.reverseBetween(&head,2,4);
 	while(h)
 	{
 		cout<<h->val<<endl;
 		h = h->next;
 	}
}

